ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -1170,5 +1183,34 @@ static Offset getIndexSegmentStartOffset(String filename) {
     }
     return new Offset(logSegmentName, Long.parseLong(startOffsetValue));
   }
+
+  /**
+   * An {@link IndexEntry} {@link Iterator} for a sealed {@link IndexSegment}. This {@link Iterator} should
+   * only be used in the compaction so that the {@link IndexSegment} should be sealed already.
+   */
+  class SealedIndexSegmentEntryIterator implements Iterator<IndexEntry> {
+    private int currentIdx = 0;
+    private ByteBuffer mmap = serEntries.duplicate();
+    private byte[] valueBuf = new byte[valueSize];
+
+    @Override
+    public boolean hasNext() {
+      return currentIdx < numberOfEntries(mmap);
+    }
+
+    @Override
+    public IndexEntry next() {
+      try {
+        StoreKey key = getKeyAt(mmap, currentIdx);
+        mmap.get(valueBuf);
+        return new IndexEntry(key, new IndexValue(startOffset.getName(), ByteBuffer.wrap(valueBuf), getVersion()));
+      } catch (Exception e) {
+        logger.error("Failed to read index entry at " + currentIdx, e);
+      } finally {
+        currentIdx++;
+      }
+      return null;
```

Comment:
This will cause NullPointException if it failed to read index. Is this intended?

Reply:
that's right, will update.